### PR TITLE
fix($parse): respect the interceptor.$stateful flag

### PR DESCRIPTION
**Pull request by jbedard**
~~This actually depends on #16009 because that (unnecessary) `$stateful` flag was previously being ignored in the case of the failing test.~~

### CODE REVIEW

**Comment by jbedard** [comment]
For the failing test (which depends on #16009): previously it only passed because the `inputsWatchDelegate` was incorrectly used due to [this block](https://github.com/angular/angular.js/blob/b5118ac6a9e0a327b31094b3fdcdc0432b23ad2f/src/ng/parse.js#L1959-L1962) assigning the `inputs` then the `inputsWatchDelegate` was [later used](https://github.com/angular/angular.js/blob/b5118ac6a9e0a327b31094b3fdcdc0432b23ad2f/src/ng/parse.js#L1890-L1891) because of those inputs.  This is incorrect because a $stateful filter/interceptor must always be invoked, not only if it's inputs change, like the added tests demonstrate.

File: src/ng/parse.js
Comment (gkalpak) [inline src/ng/parse.js]: If we remove this, what happens with "custom" `$$watchDelegates` (such as the ones added by `$interpolate`)?

Comment (jbedard) [inline src/ng/parse.js]: This actually fixes a bug because that lost `$$watchDelegate` knows nothing about the interceptor and is built specifically for that interpolate fn.

For example watching `$parse($interpolate("{{foo}}"), function() { return "bar"; })` will always output the "foo" var even though it is wrapped in an interceptor that always returns "bar", yet executing that parsed-expression will always return "bar"...

Comment (gkalpak) [inline src/ng/parse.js]: Where is the test for this fix then? 😛

Comment (gkalpak) [inline src/ng/parse.js]: FWIW, I am still confused about what happens to `$$watchDelegates`. when combined with interceptors? Are they just ignored?

Comment (jbedard) [inline src/ng/parse.js]: Yeah they are just ignored, because the interceptor might have made that `$$watchDelegate` completely invalid. However the returned ("interecepted"?) fn might have some constant/oneTime/inputs properties that might cause a `$$watchDelegate` to be added.

Comment (jbedard) [inline src/ng/parse.js]: Here's an example, I'll try to add a test for this...
```
$parse($interpolate("a{{b}}"), function(x) { return x + 'c'; })
```
`$interpolate` adds a special `$$watchDelegate` for an efficient way to watch the interpolated-expression, but that delegate of course knows nothing about the `x + 'c'` interceptor.  Previously this line we're commenting on would blindly copy the delegate creating an expression which produces the correct value when executed (with the `'c'` concatenated at the end) but the wrong value when watched due to the use of a `$$watchDelegate` that knows nothing about the `x + 'c'`...

Comment (jbedard) [inline src/ng/parse.js]: I've added a test for this: "should not propogate $$watchDelegate to the interceptor wrapped expression"

**Comment by Narretz** [comment]
Looks like a few tests failed. And there's also an eslint error

**Comment by jbedard** [comment]
Fixed the lint issue. Not sure what the test errors were (don't think they failed before), we'll see if they fail again...

File: src/ng/parse.js
Comment (gkalpak) [inline src/ng/parse.js]: I see. And we are fine with that because we never use `$$watchDelegate` with interceptors in core, right?
I might add a comment (where?) about that, i.e. the you can't use `$$watchDelegate` with interceptors.

**Comment by jbedard** [comment]
I think the remaining test failure is actually caused by #16021. I'll probably end up trying to fix both in this PR since they seem overlap a lot...

**Comment by jbedard** [comment]
@lgalfaso would be great to get your review/opinion if you have a chance

File: src/ng/parse.js
Comment (lgalfaso) [inline src/ng/parse.js]: Are you sure that `locals` should be sent to both interceptors?

File: src/ng/parse.js
Comment (lgalfaso) [inline src/ng/parse.js]: I understand why this chaining is needed, just wonder if there is a cleaner solution.

**Review by lgalfaso — COMMENTED** [review]
Overall, I think it is a good change and a nice cleanup. Just some minor comments.

File: src/ng/parse.js
Comment (jbedard) [inline src/ng/parse.js]: Good question. I'll look into it and try to add some tests...

Comment (jbedard) [inline src/ng/parse.js]: This is equivalent to today. The interceptor [is passed the scope and locals](https://github.com/angular/angular.js/blob/v1.6.5/src/ng/parse.js#L1963). Then if it is double-intercepted then it will [pass them through](https://github.com/angular/angular.js/blob/v1.6.5/src/ng/parse.js#L1962) until reaching the original expression.

I would be tempted to remove these args from the interceptor though, and only pass them to the original wrapped expression. Probably in a separate PR though...

File: src/ng/parse.js
Comment (jbedard) [inline src/ng/parse.js]: This is all internal so it can always be changed later. I'm open to any ideas though 👍

File: src/ng/parse.js
Comment (gkalpak) [inline src/ng/parse.js]: How could `parsedExpresion.inputs` be truthy and `parsedException.$$intercepted.inputs` be falsy?
Can you give an example scenario?

Comment (jbedard) [inline src/ng/parse.js]: It's basically detecting [this case](https://github.com/angular/angular.js/blob/v1.6.5/src/ng/parse.js#L1973). This is when the intercepted expression had no inputs, so the expression itself becomes the input.

For example
* `$parse("x", stringify)`: `x` has no inputs, but we don't want to execute `stringify(x)` per-digest so we set `x` itself as the input and `useInputs = true` to handle this
* `$parse("x | number", stringify)`: `x | number` already has `x` as an input so we just continue using that input (and `useInputs = false`)

Comment (jbedard) [inline src/ng/parse.js]: The fact that this line basically "detects" the `useInputs = true` from `addInterceptor` is something I don't like about this PR atm. It already existed before, but I guess now it is spread across two methods which seems worse to me. But it's all internal and could always be changed...

File: src/ng/parse.js
Comment (jbedard) [inline src/ng/parse.js]: See https://github.com/angular/angular.js/pull/16091 which might be nice to merge before this PR.

File: src/ng/parse.js
Comment (gkalpak) [inline src/ng/parse.js]: I can't say I can fully wrap my head around this, but the tests pass, so... :smiley:

**Review by gkalpak — APPROVED** [review]

File: src/ng/parse.js
Comment (jbedard) [inline src/ng/parse.js]: Which part? Anything I can explain more/better?  There's a couple weird things being changed in this PR, and I also get confused after not looking at it for a while ^_^

Comment (gkalpak) [inline src/ng/parse.js]: The `useInputs` bit still confuses me (but in tests we trust).
<sub>(At some point I need to grab my debugger and go through some usecases, but this doesn't have to hold this PR back.)</sub>

Comment (jbedard) [inline src/ng/parse.js]: The `useInputs` is also the part I don't like.

Note that I think it should really be renamed to something such as `theInputIsTheInterceptedExpression` (where normally the input is the input **to** the intercepted expression).

It is also just a performance thing, dropping all `useInputs == true` logic would still function correctly.

File: src/ng/parse.js
Comment (jbedard) [inline src/ng/parse.js]: This method is now simplified after #16091 merged.

**Pull request merged**
